src/lib/storage/index/partial_hash/flat_map_iterator.hpp

```diff
@@ -2,6 +2,24 @@
 
 #include "types.hpp"
 
+/**
+ * The choice of the concrete iterator implementation was subject to long discussion. The PartialHashIndex uses the
+ * pointer to implementation (pImpl) idom, that means  i.e. a non-templated class PartialHashIndex holds an opaque
+ * pointer to its templated implementation PartialHashIndexImpl. Since the index access is done through its iterators
+ * (more precisely through its access methods that also work with iterators), a non-templated iterator was also needed.
+ *
+ * Several attempts were made to achieve this, e.g. passing the iterator to the index as a template, keeping the
+ * templated iterators inside the IndexImpl and access them via the index's access methods. However, these attempts
+ * were not only very ugly, it was also not possible to archive the desirend effect because templatization and class
+ * polymorphism do not got well together.
```

Comment:
Can we also formulate this more consciously?

Example:
```
The following design attempts were discarded:

Iterator as a template parameter: <documentation of issues with that approach>

Returning BaseFlatMapIteratorImpl pointers rather than FlatMapIterator instances: <documentation of issues with that approach>
```

As listed above, can you also document the observed issues with returning iterator pointers, @tjjordan?

`because templatization and class polymorphism do not go well together.` Can you be more specific since templates are one option to achieve polymorphism, namely static / compile-time polymorphism